test/integration/replicaset/replicaset_test.go

```diff
@@ -272,246 +279,698 @@ func TestAdoption(t *testing.T) {
+	newSelectorLabels = map[string]string{"changed_name_apps_v1beta2": "changed_test_apps_v1beta2"}
+	rsV1beta2.Spec.Selector.MatchLabels = newSelectorLabels
+	rsV1beta2.Spec.Template.Labels = newSelectorLabels
+	_, err = clientSet.AppsV1beta2().ReplicaSets(ns.Name).Update(rsV1beta2)
+	if err == nil {
+		t.Fatalf("failed to provide validation error when changing immutable selector when updating apps/v1beta2 replicaset %s", rsV1beta2.Name)
+	}
+	expectedErrType := "Invalid value"
+	expectedErrDetail := "field is immutable"
+	if !strings.Contains(err.Error(), expectedErrType) || !strings.Contains(err.Error(), expectedErrDetail) {
+		t.Errorf("error message does not match, expected type: %s, expected detail: %s, got: %s", expectedErrType, expectedErrDetail, err.Error())
+	}
+}
+
+// Update .Spec.Replicas to replicas and verify .Status.Replicas is changed accordingly
+func scaleRS(t *testing.T, c clientset.Interface, rs *v1beta1.ReplicaSet, replicas int32) {
+	rsClient := c.Extensions().ReplicaSets(rs.Namespace)
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newRS, err := rsClient.Get(rs.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		*newRS.Spec.Replicas = replicas
+		rs, err = rsClient.Update(newRS)
+		return err
+	}); err != nil {
+		t.Fatalf("Failed to update .Spec.Replicas to %d for rs %s: %v", replicas, rs.Name, err)
 	}
-	t.Logf("patched rs = %#v", rs)
-	// wait for the rs select both pods and delete one of them
-	if err := wait.Poll(10*time.Second, 60*time.Second, func() (bool, error) {
-		return verifyRemainingObjects(t, clientSet, ns.Name, 1, 1)
+	waitRSStable(t, c, rs)
+}
+
+func updatePod(t *testing.T, podClient typedv1.PodInterface, pod *v1.Pod, updateFunc func(*v1.Pod)) {
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newPod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		updateFunc(newPod)
+		_, err = podClient.Update(newPod)
+		return err
 	}); err != nil {
-		t.Fatal(err)
+		t.Fatalf("Failed to update pod %s: %v", pod.Name, err)
 	}
-	close(stopCh)
 }
 
-func TestUpdateSelectorToRemoveControllerRef(t *testing.T) {
-	// We have pod1, pod2 and rs. rs.spec.replicas=2. At first rs.Selector
-	// matches pod1 and pod2; change the selector to match only pod1. Verify
-	// that rs creates one more pod, so there are 3 pods. Also verify that
-	// pod2's controllerRef is cleared.
-	s, closeFn, rm, informers, clientSet := rmSetup(t)
+func getPods(t *testing.T, podClient typedv1.PodInterface, labelMap map[string]string) *v1.PodList {
+	podSelector := labels.Set(labelMap).AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+	pods, err := podClient.List(options)
+	if err != nil {
+		t.Fatalf("Failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
+	}
+	if pods == nil {
+		t.Fatalf("Obtained a nil list of pods")
+	}
+	return pods
+}
+
+func TestReplicaSetBasics(t *testing.T) {
```

Comment:
@enisoc I refactored the deleting/failed test a lot, but it still doesn't address 2 things:
(1) failed pod should stay around even scaling down RS to zero
(2) new pod replacing deleting pod does not get created upfront until scaling up happens

I am investigating (2) via pooling, but i think it may be wise for you to review (just) this test to ensure i get the direction right.

Reply:
To sum up, i am able to verify:
(1) deleting pod and failed pod survive initially
(2) deleting pod survives when RS scales down to zero (and also after RS scales back up)
(3) deleting pod's deletion timestamp does not get removed

i am not able to verify:
(1) failed pod survives when RS scales down to zero (and also after RS scales back up)
(2) the 2 new pods are created immediately (creation only happens after scaling back up)

Reply:
I recalled separating the deleting-failed test and scaling test before.
Without scaling, 2 new pods are not created. Failed pod does not survive too.